File: ios/Classes/WxbwechatPlugin.m
#import "WxbwechatPlugin.h"

@implementation WxbwechatPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"wxbwechat"
            binaryMessenger:[registrar messenger]];
  WxbwechatPlugin* instance = [[WxbwechatPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  } else if ([@"register" isEqualToString:call.method]){
      [WXApi registerApp:@"wx7df4c6aef7dd5845"];
      result(@"注册微信开发者");
  }else if ([@"send" isEqualToString:call.method]){
      NSString *name = call.arguments[@"userName"];
      NSString *cardId = call.arguments[@"cardId"];
      NSString *imageUrl = call.arguments[@"headimgurl"];
      NSString *type = call.arguments[@"type"];
      
      [self shareToWechatWithName:name visitingCardId:cardId imgUrl:imageUrl type:type];
  }else if ([@"shareweb" isEqualToString:call.method]){
      NSString *title = call.arguments[@"title"];
      NSString *desc = call.arguments[@"desc"];
      NSString *url = call.arguments[@"url"];
      NSString *thumbUrl = call.arguments[@"thumbUrl"];
      NSString *type = call.arguments[@"type"];
      [self shareWebToWxWithTitle:title desc:desc thumbImage:thumbUrl url:url type:type];
  } else {
    result(FlutterMethodNotImplemented);
  }
}

-(void)shareWebToWxWithTitle:(NSString *)title desc:(NSString *)desc thumbImage:(NSString *)thumbImg url:(NSString *)url type:(NSString *)type{
    WXWebpageObject *webpageObject = [WXWebpageObject object];
    webpageObject.webpageUrl = url;
    
    WXMediaMessage *message = [WXMediaMessage message];
    message.title = title;
    message.description = desc;
    NSData *data = [NSData dataWithContentsOfURL:[NSURL URLWithString:thumbImg]];
    [message setThumbImage:[UIImage imageWithData:data]];
    message.mediaObject = webpageObject;
    
    SendMessageToWXReq *req = [[SendMessageToWXReq alloc] init];
    req.bText = NO;
    req.message = message;
    req.scene = [type isEqualToString:@"moment"] ? WXSceneTimeline : WXSceneSession;
    [WXApi sendReq:req];
}

- (void)shareToWechatWithName:(NSString *)name visitingCardId:(NSString *)cardId imgUrl:(NSString *)imgUrl type:(NSString *)type {
    NSString *title = [NSString stringWithFormat:@"你好，我是%@，这是我的名片",name];
    NSString *path = [NSString stringWithFormat:@"pages/personal_card/card?visitingCardId=%@",cardId];
    WXMiniProgramObject *wxMiniObject = [[WXMiniProgramObject alloc]init];
    wxMiniObject.webpageUrl = @"https://api-test-c.wabgxiaobao.co/visiting-card/error";
    wxMiniObject.userName = @"gh_d2b176e76ef5";
    wxMiniObject.path = path;
    NSData *data = [NSData dataWithContentsOfURL:[NSURL URLWithString:imgUrl]];
    UIImage *img = [UIImage imageWithData:data];
    wxMiniObject.hdImageData = [self reSizeImageData:img maxImageSize:500 maxFileSizeWithKB:100];
    wxMiniObject.miniProgramType = [type isEqualToString:@"release"] ? WXMiniProgramTypeRelease : WXMiniProgramTypePreview;
    wxMiniObject.withShareTicket = YES;
    
    WXMediaMessage *message = [[WXMediaMessage alloc]init];
    message.title = title;
    message.description = title;
    message.mediaObject = wxMiniObject;
    message.thumbData = nil;
    
    SendMessageToWXReq *req = [[SendMessageToWXReq alloc]init];
    req.message = message;
    req.bText = NO;
    req.scene = WXSceneSession;
    [WXApi sendReq:req];
}


- (NSData *)reSizeImageData:(UIImage *)sourceImage maxImageSize:(CGFloat)maxImageSize maxFileSizeWithKB:(CGFloat)maxFileSize
{
    
    if (maxFileSize <= 0.0)  maxFileSize = 1024.0;
    if (maxImageSize <= 0.0) maxImageSize = 1024.0;
    
    //先调整分辨率
    CGSize newSize = CGSizeMake(sourceImage.size.width, sourceImage.size.height);
    
    CGFloat tempHeight = newSize.height / maxImageSize;
    CGFloat tempWidth = newSize.width / maxImageSize;
    
    if (tempWidth > 1.0 && tempWidth > tempHeight) {
        newSize = CGSizeMake(sourceImage.size.width / tempWidth, sourceImage.size.height / tempWidth);
    }
    else if (tempHeight > 1.0 && tempWidth < tempHeight){
        newSize = CGSizeMake(sourceImage.size.width / tempHeight, sourceImage.size.height / tempHeight);
    }
    
    UIGraphicsBeginImageContext(newSize);
    [sourceImage drawInRect:CGRectMake(0,0,newSize.width,newSize.height)];
    UIImage *newImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    //调整大小
    NSData *imageData = UIImageJPEGRepresentation(newImage,1.0);
    CGFloat sizeOriginKB = imageData.length / 1024.0;
    
    CGFloat resizeRate = 0.9;
    while (sizeOriginKB > maxFileSize && resizeRate > 0.1) {
        imageData = UIImageJPEGRepresentation(newImage,resizeRate);
        sizeOriginKB = imageData.length / 1024.0;
        resizeRate -= 0.1;
    }
    
    return imageData;
}


@end

File: android/src/main/java/com/example/wxbwechat/WxbwechatPlugin.java
package com.example.wxbwechat;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

import com.tencent.mm.opensdk.modelmsg.WXWebpageObject;
import com.tencent.mm.opensdk.openapi.IWXAPI;
import com.tencent.mm.opensdk.openapi.IWXAPIEventHandler;
import com.tencent.mm.opensdk.openapi.WXAPIFactory;
import com.tencent.mm.opensdk.modelbiz.WXLaunchMiniProgram;
import com.tencent.mm.opensdk.modelmsg.SendMessageToWX;
import com.tencent.mm.opensdk.modelmsg.WXMediaMessage;
import com.tencent.mm.opensdk.modelmsg.WXMiniProgramObject;


import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import android.os.Message;
import android.os.Handler;

/** WxbwechatPlugin */
public class WxbwechatPlugin implements MethodCallHandler {

  private static IWXAPI api;
  private Context context;
  private WXMediaMessage message;
  private Bitmap bitmap;
  private String kind;

  private WxbwechatPlugin(Context ctx) {
    context = ctx;
  }

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "wxbwechat");
    channel.setMethodCallHandler(new WxbwechatPlugin(registrar.context()));
//    channel.setMethodCallHandler(new WxbwechatPlugin());
  }



  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else if(call.method.equals("register")) {
      api = WXAPIFactory.createWXAPI(context,"wx7df4c6aef7dd5845",true);
      api.registerApp("wx7df4c6aef7dd5845");
      result.success("");
    } else if (call.method.equals("send")) {
      String name = call.argument("userName");
      String cardId = call.argument("cardId");
      String headimgurl = call.argument("headimgurl");
      String type = call.argument("type");

      sendCard(name,cardId,headimgurl,type);
    }else if (call.method.equals("shareweb")){
      String title = call.argument("title");
      String desc = call.argument("desc");
      String thumbUrl = call.argument("thumbUrl");
      String url = call.argument("url");
      String type = call.argument("type");
      shareWebToWx(title,desc,thumbUrl,url,type);
    } else{
      result.notImplemented();
    }
  }

  private Handler handler = new Handler(new Handler.Callback() {
    @Override
    public boolean handleMessage(Message osMessage) {

      SendMessageToWX.Req req = new SendMessageToWX.Req();
      int maxSize = osMessage.what==0 ? 128 : 32;

      if (bitmap != null) {
        Bitmap thumbBitmap = Bitmap.createScaledBitmap(bitmap, 200, 200, true);
        message.thumbData = bitmap2Bytes(bitmap,maxSize);
      }

      if (osMessage.what==0) {
        req.transaction = "miniProgram" + System.currentTimeMillis();

        req.scene = SendMessageToWX.Req.WXSceneSession;  // 目前只支持会话
        req.message = message;
        api.sendReq(req);
      }else{
        req.transaction = String.valueOf(System.currentTimeMillis());
        req.message =message;
        req.scene = kind.equals("moment") ? SendMessageToWX.Req.WXSceneTimeline : SendMessageToWX.Req.WXSceneSession;;
        api.sendReq(req);
      }


      return false;
    }
  });

  private void shareWebToWx(String title, String desc, final String thumbUrl, String url, String type) {
    //初始化一个WXWebpageObject，填写url
    kind = type;
    WXWebpageObject webpage = new WXWebpageObject();
    webpage.webpageUrl =url;

    message = new WXMediaMessage(webpage);
    message.title =title;
    message.description =desc;

    new Thread() {
      public void run() {
        Message osMessage = new Message();
        bitmap = GetBitmap(thumbUrl);
        osMessage.what = 1;
        handler.sendMessage(osMessage);
      }
    }.start();

//构造一个Req

  }

  private void sendCard(String name, String cardId, final String headImgurl, String type) {
    WXMiniProgramObject miniProgramObj = new WXMiniProgramObject();
    miniProgramObj.userName = "gh_d2b176e76ef5";// 小程序原始id
    miniProgramObj.webpageUrl = "https://api-test-c.wabgxiaobao.co/visiting-card/error"; // 兼容低版本的网页链接
    miniProgramObj.miniprogramType = type.equals("release") ? WXMiniProgramObject.MINIPTOGRAM_TYPE_RELEASE : WXMiniProgramObject.MINIPROGRAM_TYPE_PREVIEW; // 正式版:0，测试版:1，体验版:2

    String path = "pages/personal_card/card?visitingCardId=".concat(cardId);
    miniProgramObj.path = path; // 小程序页面路径

    message = new WXMediaMessage(miniProgramObj);
    String title = "你好，我是"+name+",这是我的名片";
    message.title = title; // 小程序消息title
    message.description = title; // 小程序消息desc
    new Thread() {
      public void run() {
        Message osMessage = new Message();
        bitmap = GetBitmap(headImgurl);
        osMessage.what = 0;
        handler.sendMessage(osMessage);
      }
    }.start();

  }

  /**
   * Bitmap转换成byte[]并且进行压缩,压缩到不大于maxkb
   * @param bitmap
   * @param
   * @return
   */
  public static byte[] bitmap2Bytes(Bitmap bitmap, int maxkb) {
    ByteArrayOutputStream output = new ByteArrayOutputStream();
    bitmap.compress(Bitmap.CompressFormat.PNG, 100, output);
    int options = 100;
    while (output.toByteArray().length > maxkb&& options != 10) {
      output.reset(); //清空output
      bitmap.compress(Bitmap.CompressFormat.JPEG, options, output);//这里压缩options%，把压缩后的数据存放到output中
      options -= 10;
    }
    return output.toByteArray();
  }


  public Bitmap GetBitmap(String url) {
    Bitmap bitmap = null;
    InputStream in = null;
    BufferedOutputStream out = null;
    try {
      in = new BufferedInputStream(new URL(url).openStream(), 1024);
      final ByteArrayOutputStream dataStream = new ByteArrayOutputStream();
      out = new BufferedOutputStream(dataStream, 1024);
      copy(in, out);
      out.flush();
      byte[] data = dataStream.toByteArray();
      bitmap = BitmapFactory.decodeByteArray(data, 0, data.length);
      return bitmap;
    } catch (IOException e) {
      e.printStackTrace();
      return null;
    }
  }

  private static void copy(InputStream in, OutputStream out) throws IOException {
    byte[] b = new byte[1024];
    int read;
    while ((read = in.read(b)) != -1) {
      out.write(b, 0, read);
    }
  }
}
